File: ios/Framework/Core/Base/Interactor.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit


@objc public class Interactor: NSObject {

	public var actionName: String?

	public var onSuccess: (Void -> Void)?
	public var onFailure: (NSError -> Void)?

	public var lastError: NSError?

	public let screenlet: BaseScreenlet?

	internal var cancelled = false


	public init(screenlet: BaseScreenlet?) {
		self.screenlet = screenlet

		super.init()
	}

	override public convenience init() {
		self.init(screenlet: nil)
	}

	public func callOnSuccess() {
		if !cancelled {
			dispatch_main {
				self.onSuccess?()
				self.finishWithError(nil)
			}
		}
	}

	public func callOnFailure(error: NSError) {
		if !cancelled {
			dispatch_main {
				self.onFailure?(error)
				self.finishWithError(error)
			}
		}
	}

	public func start() -> Bool {
		cancelled = false
		return false
	}

	public func cancel() {
		callOnFailure(NSError.errorWithCause(.Cancelled))
		cancelled = true
	}

	public func interactionResult() -> AnyObject? {
		return nil
	}

	private func finishWithError(error: NSError?) {
		screenlet?.endInteractor(self, error: error)

		// break retain cycle
		onSuccess = nil
		onFailure = nil
	}

}

File: xamarin/Framework/LiferayScreens.iOS/Base/Interactor.cs
using System;
using ObjCRuntime;
using Foundation;

namespace LiferayScreens
{
    // @interface Interactor : NSObject
    [BaseType(typeof(NSObject))]
    interface Interactor
    {
        // @property (copy, nonatomic) NSString * _Nullable actionName;
        [NullAllowed, Export("actionName")]
        string ActionName { get; set; }

        // @property (copy, nonatomic) void (^ _Nullable)(void) onSuccess;
        [NullAllowed, Export("onSuccess", ArgumentSemantic.Copy)]
        Action OnSuccess { get; set; }

        // @property (copy, nonatomic) void (^ _Nullable)(NSError * _Nonnull) onFailure;
        [NullAllowed, Export("onFailure", ArgumentSemantic.Copy)]
        Action<NSError> OnFailure { get; set; }

        // @property (nonatomic, strong) NSError * _Nullable lastError;
        [NullAllowed, Export("lastError", ArgumentSemantic.Strong)]
        NSError LastError { get; set; }

        // @property (readonly, nonatomic, strong) BaseScreenlet * _Nullable screenlet;
        [NullAllowed, Export("screenlet", ArgumentSemantic.Strong)]
        BaseScreenlet Screenlet { get; }

        // -(instancetype _Nonnull)initWithScreenlet:(BaseScreenlet * _Nullable)screenlet __attribute__((objc_designated_initializer));
        [Export("initWithScreenlet:")]
        [DesignatedInitializer]
        IntPtr Constructor([NullAllowed] BaseScreenlet screenlet);

        // -(void)callOnSuccess;
        [Export("callOnSuccess")]
        void CallOnSuccess();

        // -(void)callOnFailure:(NSError * _Nonnull)error;
        [Export("callOnFailure:")]
        void CallOnFailure(NSError error);

        // -(BOOL)start __attribute__((warn_unused_result));
        [Export("start")]
        bool Start();

        // -(void)cancel;
        [Export("cancel")]
        void Cancel();

        // -(id _Nullable)interactionResult __attribute__((warn_unused_result));
        [NullAllowed, Export("interactionResult")]
        NSObject InteractionResult();
    }
}
